File: backend/app/schemas/job.py
from datetime import datetime
from enum import Enum

from pydantic import BaseModel, ConfigDict, Field


class JobStatus(str, Enum):
    INTERESTED = "interested"
    APPLIED = "applied"
    INTERVIEW = "interview"
    REJECTED = "rejected"
    OFFER = "offer"
    GHOSTED = "ghosted"


class JobParseRequest(BaseModel):
    description: str = Field(min_length=20, max_length=50000)


class JobParsed(BaseModel):
    company: str = Field(min_length=1, max_length=255)
    role: str = Field(min_length=1, max_length=255)
    requirements: list[str] = Field(default_factory=list)
    skills: list[str] = Field(default_factory=list)
    responsibilities: list[str] = Field(default_factory=list)


class JobCreate(BaseModel):
    company: str = Field(min_length=1, max_length=255)
    role: str = Field(min_length=1, max_length=255)
    raw_description: str = Field(min_length=20)
    status: JobStatus = JobStatus.INTERESTED
    requirements: list[str] = Field(default_factory=list)
    skills: list[str] = Field(default_factory=list)
    responsibilities: list[str] = Field(default_factory=list)
    job_url: str | None = None


class JobUpdate(BaseModel):
    company: str | None = None
    role: str | None = None
    status: JobStatus | None = None
    raw_description: str | None = None
    requirements: list[str] | None = None
    skills: list[str] | None = None
    responsibilities: list[str] | None = None
    job_url: str | None = None


class JobRead(BaseModel):
    id: int
    company: str
    role: str
    status: JobStatus
    raw_description: str
    requirements: list[str]
    skills: list[str]
    responsibilities: list[str]
    job_url: str | None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/types/job.ts
export type JobStatus =
  | "interested"
  | "applied"
  | "interview"
  | "rejected"
  | "offer"
  | "ghosted";

export type Job = {
  id: number;
  company: string;
  role: string;
  status: JobStatus;
  raw_description: string;
  requirements: string[];
  skills: string[];
  responsibilities: string[];
  job_url: string | null;
  created_at: string;
  updated_at: string;
};

export type JobParsed = {
  company: string;
  role: string;
  requirements: string[];
  skills: string[];
  responsibilities: string[];
};

export type JobCreate = {
  company: string;
  role: string;
  raw_description: string;
  status?: JobStatus;
  requirements?: string[];
  skills?: string[];
  responsibilities?: string[];
  job_url?: string | null;
};

export type JobUpdate = {
  company?: string;
  role?: string;
  status?: JobStatus;
  raw_description?: string;
  requirements?: string[];
  skills?: string[];
  responsibilities?: string[];
  job_url?: string | null;
};

export const JOB_STATUSES: JobStatus[] = [
  "interested",
  "applied",
  "interview",
  "rejected",
  "offer",
  "ghosted",
];

export const JOB_STATUS_LABELS: Record<JobStatus, string> = {
  interested: "Interested",
  applied: "Applied",
  interview: "Interview",
  rejected: "Rejected",
  offer: "Offer",
  ghosted: "Ghosted",
};

export function formatJobDate(iso: string): string {
  return new Date(iso).toLocaleString(undefined, {
    dateStyle: "medium",
    timeStyle: "short",
  });
}
